lib/invoker/power/balancer.rb

```diff
@@ -43,29 +43,47 @@ def initialize(connection, protocol)
         @connection = connection
         @protocol = protocol
         @http_parser = HttpParser.new(protocol)
+        @response_http_parser = HttpParser.new(protocol)
         @session = nil
         @buffer = []
       end
 
       def install_callbacks
+        http_parser.on_url { |url| url_received(url) }
         http_parser.on_headers_complete { |headers| headers_received(headers) }
         http_parser.on_message_complete { |full_message| complete_message_received(full_message) }
+
+        @response_http_parser.on_message_complete { response_complete }
+
         connection.on_data { |data| upstream_data(data) }
         connection.on_response { |backend, data| backend_data(backend, data) }
         connection.on_finish { |backend, name| frontend_disconnect(backend, name) }
       end
 
+      def response_complete
+        # disconnect from backend after each response, since subsequent
+        # requests from persistent HTTP connections might be proxied to a stale
+        # backend
+        EventMachine.next_tick do # still needs chance to send data
+          connection.unbind_backend(@session)
+          @session = nil
+          @response_http_parser.reset
+          http_parser.reset
+        end
+      end
```

Comment:
This is actually required since the previously, there was a one-to-one mapping between incoming domain and backend. Now, since a single incoming domain could have multiple backends depending on the path, we need to make sure that the backend is not persisted across the same request coming in from the same domain.